File: wetty-chat-mobile/src/api/friends.ts
import apiClient from './client';
import type { MemberSummary } from './users';

export type FriendRequestStatus = 'pending' | 'accepted' | 'rejected' | 'archived';

/**
 * How a user gates incoming friend requests.
 * - `direct`: anyone may request; no verification message allowed (switch off).
 * - `need_message`: requester must attach a non-empty verification message.
 * - `forbid`: all incoming requests are rejected.
 * - `question`: requester must answer a pre-set question (manual review).
 */
export type FriendAddVerificationMode = 'direct' | 'need_message' | 'forbid' | 'question';

export interface FriendResponse {
  user: MemberSummary;
  since: string;
}

export interface ListFriendsResponse {
  friends: FriendResponse[];
}

/** The authenticated user's current relationship with one profile user. */
export interface FriendRelationshipResponse {
  peerUid: number;
  isFriend: boolean;
  dmChatId: string | null;
  blocking: boolean;
  blockedBy: boolean;
  canDm: boolean;
  hasPendingOutgoingRequest: boolean;
}

export interface FriendRequestResponse {
  id: string;
  from: MemberSummary;
  to: MemberSummary;
  status: FriendRequestStatus;
  createdAt: string;
  decidedAt: string | null;
  /** Verification message (need_message) or the requester's answer (question); null for direct. */
  message?: string | null;
  /** Snapshot of the target's question (question mode only); null otherwise. */
  question?: string | null;
}

export type FriendRequestDirection = 'incoming' | 'outgoing';

export interface FriendRequestHistoryEntry extends FriendRequestResponse {
  /** Direction relative to the current user, as decided by the server. */
  direction: FriendRequestDirection;
}

export interface ListFriendRequestHistoryResponse {
  requests: FriendRequestHistoryEntry[];
}

export interface FriendSettingsResponse {
  mode: FriendAddVerificationMode;
  question: string | null;
}

export interface FriendAddInfoResponse {
  mode: FriendAddVerificationMode;
  question: string | null;
}

export interface UpdateFriendSettingsBody {
  mode: FriendAddVerificationMode;
  question: string | null;
}

export const friendsApi = {
  listFriends: async (): Promise<FriendResponse[]> => {
    const res = await apiClient.get<ListFriendsResponse>('/friends');
    return res.data.friends;
  },

  getRelationship: async (uid: number): Promise<FriendRelationshipResponse> => {
    const res = await apiClient.get<FriendRelationshipResponse>(`/friends/${uid}`);
    return res.data;
  },

  removeFriend: async (uid: number): Promise<void> => {
    await apiClient.delete(`/friends/${uid}`);
  },

  /**
   * Send a friend request. `message` is the verification message (need_message)
   * or the answer to the target's question (question); omitted for direct.
   */
  createRequest: async (toUid: number, message?: string): Promise<FriendRequestResponse> => {
    const res = await apiClient.post<FriendRequestResponse>('/friends/requests', { toUid, message });
    return res.data;
  },

  /** Friend-request history in server-defined order, optionally filtered by archive view. */
  listRequestHistory: async (archived?: boolean): Promise<FriendRequestHistoryEntry[]> => {
    const res = await apiClient.get<ListFriendRequestHistoryResponse>('/friends/requests', { params: { archived } });
    return res.data.requests;
  },

  acceptRequest: async (requestId: string): Promise<FriendRequestResponse> => {
    const res = await apiClient.post<FriendRequestResponse>(`/friends/requests/${requestId}/accept`);
    return res.data;
  },

  rejectRequest: async (requestId: string): Promise<FriendRequestResponse> => {
    const res = await apiClient.post<FriendRequestResponse>(`/friends/requests/${requestId}/reject`);
    return res.data;
  },

  archiveRequest: async (requestId: string): Promise<void> => {
    await apiClient.put(`/friends/requests/${requestId}/archive`);
  },

  getMySettings: async (): Promise<FriendSettingsResponse> => {
    const res = await apiClient.get<FriendSettingsResponse>('/friends/me/settings');
    return res.data;
  },

  updateMySettings: async (body: UpdateFriendSettingsBody): Promise<FriendSettingsResponse> => {
    const res = await apiClient.put<FriendSettingsResponse>('/friends/me/settings', body);
    return res.data;
  },

  /** What a requester needs in order to add `uid` as a friend. */
  getAddInfo: async (uid: number): Promise<FriendAddInfoResponse> => {
    const res = await apiClient.get<FriendAddInfoResponse>(`/friends/add-info/${uid}`);
    return res.data;
  },
};

File: backend/src/dto/friends.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::{
    dto::users::MemberSummary,
    models::{FriendAddVerificationMode, FriendRequestStatus},
};

#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct FriendResponse {
    pub user: MemberSummary,
    pub since: DateTime<Utc>,
}

/// The current user's relationship with one profile user.
#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct FriendRelationshipResponse {
    pub peer_uid: i32,
    pub is_friend: bool,
    /// The canonical DM group ID, when one exists for this relationship.
    #[serde(with = "crate::serde_i64_string::opt")]
    #[schema(value_type = Option<String>)]
    pub dm_chat_id: Option<i64>,
    /// The current user has blocked `peerUid`.
    pub blocking: bool,
    /// `peerUid` has blocked the current user.
    pub blocked_by: bool,
    /// The server-authoritative decision for whether the current user can send this peer a DM.
    pub can_dm: bool,
    /// The current user has an unresolved request to `peerUid`.
    pub has_pending_outgoing_request: bool,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ListFriendsResponse {
    pub friends: Vec<FriendResponse>,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct FriendRequestResponse {
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub id: i64,
    pub from: MemberSummary,
    pub to: MemberSummary,
    pub status: FriendRequestStatus,
    pub created_at: DateTime<Utc>,
    pub decided_at: Option<DateTime<Utc>>,
    /// Verification message (mode 1) or the requester's answer (mode 3); absent for direct.
    pub message: Option<String>,
    /// Snapshot of the target's question (mode 3 only); absent otherwise.
    pub question: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub enum FriendRequestDirection {
    Incoming,
    Outgoing,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct FriendRequestHistoryEntry {
    #[serde(flatten)]
    #[schema(inline)]
    pub request: FriendRequestResponse,
    /// Direction relative to the calling user.
    pub direction: FriendRequestDirection,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ListFriendRequestHistoryResponse {
    pub requests: Vec<FriendRequestHistoryEntry>,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PendingFriendRequestCountResponse {
    pub pending_incoming_count: i64,
}

/// A user's own friend-acceptance settings (GET /me/friend-settings).
#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct FriendSettingsResponse {
    pub mode: FriendAddVerificationMode,
    pub question: Option<String>,
}

/// Body for PUT /me/friend-settings.
#[derive(Debug, Clone, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct UpdateFriendSettingsBody {
    pub mode: FriendAddVerificationMode,
    #[schema(max_length = 100)]
    pub question: Option<String>,
}

/// What a requester needs to know to render the add-friend UI for a target
/// (GET /users/:uid/friend-add-info).
#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct FriendAddInfoResponse {
    pub mode: FriendAddVerificationMode,
    /// Present only when mode is `question`.
    pub question: Option<String>,
}
